RPC API compatibility test

piersy opened the pull request:
Adds a test that is segregated from normal compilation by a build tag which compares the RPC API outputs of a celo-blockchain node and an op-geth node using a migrated version of the datadir used by the celo node.

The intent is to ensure that our cel2 nodes will return rpc api responses for historical data that are consistent with the celo-blockchain node.

Run it like this:
```
go test -v ./compat_test -tags compat_test -celo-url `<celo rpc url>`  -op-geth-url `<op-geth rpc url>`
```

The build tag disables static analysis in vscode so if you still want that you need to add the build tag to your preferences:
```
{
  ...
  "go.buildTags": "compat_test",
  ...
}
```

Closes https://github.com/celo-org/celo-blockchain-planning/issues/363

piersy committed:
Add rpc api compatibility test

piersy committed:
Add license for copied code

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+}
+
+// Connects to a celo-blockchain and an op-geth node over rpc, selects the lowest head block and iterates from 0 to the
+// lowest head block comparing all blcoks, transactions receipts and logs.
```
```suggestion
// lowest head block comparing all blocks, transactions receipts and logs.
```

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	g.Go(func() error {
+		// Set up contiguous block tracking
+		latestContiguousBlock := startBlock
+		if startBlock > 0 {
+			latestContiguousBlock = startBlock - 1
+		}
```
This case only happens when modifying the code right? With the current code it is always `startBlock := uint64(0)`.

piersy replied:
Thanks, I was meaning to add a flag to set the start block, but forgot, so I will add that.

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+			allLogs = append(allLogs, logs...)
+		}
+
+		defer b.logFetchContextCancel()
```
I don't see the benefit of this, since we wait on the errgroup in the next line. So I don't think there is anything left to cancel.

piersy replied:
Yep, I think you're right here. Will remove

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	return b.remaining == 0, nil
+}
+
+// Holds results retreived from the celo and op-geth clients rpc api, fields named raw were retrieved via a direcet rpc
```
```suggestion
// Holds results retrieved from the celo and op-geth clients rpc api, fields named raw were retrieved via a direct rpc
```

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	// We also need to take into account yparity (only set on op transactions which can cause different representations for the v value)
+	// We overcome this by re-setting the v value of the celo txs to the op txs v value
+	for _, tx := range r.celoTxs {
+		// This sets the y value to be a big number where abs is nul rather than a zero length array if the numebr is zero.
```
```suggestion
		// This sets the y value to be a big number where abs is nul rather than a zero length array if the number is zero.
```

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	for _, tx := range r.celoTxs {
+		// This sets the y value to be a big number where abs is nul rather than a zero length array if the numebr is zero.
+		// It doesn't change the number but it does change the representation.
+		types.SetYNullStyleBigIfZero(tx)
```
Somehow I don't see where `SetYNullStyleBigIfZero` is defined.

piersy replied:
My mistake, when I enabled the build tag it disabled static analysis for the test, meaning when I checked for usages of `SetYNullStyleBigIfZero` I found none, when in fact the test was calling it. Will reinstate.

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+		return err
+	}
+
+	for _, opReceipt := range r.opRawReceipts {
+		if opReceipt["cumulativeGasUsed"] == 0 {
+			return fmt.Errorf("receipt with hash %q has CumulativeGasUsed == 0", opReceipt["transactionHash"])
+		}
+	}
```
Is the a story behind this check? It looks oddly specific.

piersy replied:
Yep, at one point I thought I discovered a receipt with a 0 cumulative gas used, but probably I was mistaken because on re-runs I never found one, so this can be removed.

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	})
+
+	// For each transaction in blockByNumber we retrieve it using the
+	// cethClient, ethClient, cClient, and client. Each transaction is
```
```suggestion
	// celoEthclient, opEthclient, celoClient, and opClient. Each transaction is
```

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	// not linked to a transaction we have to call the endpoint to find out if
+	// one exists.
+	if method != "eth_getBlockReceipt" {
+		if result == "nil" {
```
Should this really be `"nil"` string and not just `nil`?

piersy replied:
So actually it should be "null" which is what is returned in a json response from the RPC API if the method returns nil.

Also looking at the code for eth_getBlockReceipt I can't see a situation (apart from errors) where eth_getBlockReceipt would return "null". So I'll remove the exception for eth_getBlockReceipt and see if any issues come up on another run.

karlb commented on compat_test/compat_test.go:
```diff
@@ -0,0 +1,1012 @@
+	//it's not set,even ethcompatible ones, this is confusing so we have removed this in the op code, so we need to make
+	//sure the celo side matches.
+	delete(tx, "gatewayFee")
+	// Transaction at block 9476837 returns a nullGasPrice on celo but the op-geth side returns the gasFeeCap
+	// delete(tx, "gasPrice")
```
So does the script fail with this when run, or is this comment outdated?

piersy replied:
Outdated, thanks!

karlb reviewed (commented):
Looks good to me. Just some typo fixes and thoughts.

I skipped objdiff during the review.

piersy committed:
Update compat_test/compat_test.go

Co-authored-by: Karl Bartel <karl@karl.berlin>

piersy committed:
Re-add SetYNullStyleBigIfZero

piersy committed:
Remove check for zero cumulativeGasUsed

piersy committed:
Update compat_test/compat_test.go

Co-authored-by: Karl Bartel <karl@karl.berlin>

piersy committed:
Add start-block flag

piersy committed:
Update compat_test/compat_test.go

Co-authored-by: Karl Bartel <karl@karl.berlin>

piersy committed:
Update compat_test/compat_test.go

Co-authored-by: Karl Bartel <karl@karl.berlin>

piersy committed:
Remove old comment

piersy committed:
Add buildtag to objdiff

piersy committed:
Remove unnecessary deferred context cancel

piersy committed:
Correct check for empty objects from rpc api

piersy committed:
Merge branch 'celo8' into piersy/rpc-api-compatibility-test

piersy committed:
Handle comparing atomic pointers

introduced in a851e39cbecf116ef2dc64f0b37b0300dc762931

This is pretty sketchy but there doesn't seem to be a nice way to do
this, there's an open issue for this issue right now.

https://github.com/google/go-cmp/issues/326

piersy commented:
@karlb, unfortunately celo8 pulled in a change which has broken the test.

It's the use of new atomic types such as `atomic.Pointer` inside blocks and transactions. Atomic pointer values that share the same underlying values are not considered equal by deep equal. I tried adding some admittedly hacky workaround in the EqualObjects method, but it's not working for all cases. You can see some discussion on how to handle atomic.Pointer here https://github.com/google/go-cmp/issues/326

My next approach would be to write some manual comparison functions for transactions blocks and headers, unless you have any better ideas.

karlb commented:
> My next approach would be to write some manual comparison functions for transactions blocks and headers, unless you have any better ideas.

Disappointingly, I don't see a better way to do it. Explicit comparison functions seem to be the way to go.

piersy committed:
Add manual comparison functions

piersy committed:
Merge remote-tracking branch 'origin/celo8' into piersy/rpc-api-compatibility-test

piersy committed:
Clean up unused regexes

piersy commented:
Hi @karlb I've updated the test now to use manual comparison functions for blocks and transactions. Also I added a "fix" for json unmarshaling transactions received from the celo node. That "fix" can be removed once we have done all the migrations we need to do.

piersy committed:
Merge remote-tracking branch 'origin/celo8' into piersy/rpc-api-compatibility-test

piersy committed:
Json decode celo legacy transactions from celo node

We need to be able to support json decoding celo legacy transactions
from a celo node to allow the comparison test to fetch and compare
transactions from both celo and op-geth nodes. Unfortunately due to C
compilation issues it's not feasible to import and use the celo
ethclient.Client for this.

piersy committed:
Merge branch 'celo8' into piersy/rpc-api-compatibility-test

piersy committed:
Merge branch 'celo8' into piersy/rpc-api-compatibility-test

piersy committed:
Merge branch 'celo8' into piersy/rpc-api-compatibility-test
